src/sentry/tasks/auth.py

```diff
@@ -57,47 +59,70 @@ def email_unlink_notifications(org_id, actor_id, provider_key):
-
-    try:
-        member.remove_user()
-        member.save()
-    except (AssertionError, IntegrityError):
-        logger.warning("Could not remove 2FA noncompliant user from org", extra=logging_data)
-    else:
-        logger.info("2FA noncompliant user removed from org", extra=logging_data)
-        AuditLogEntry.objects.create(
-            actor=actor,
-            actor_key=actor_key,
-            ip_address=ip_address,
-            event=AuditLogEntryEvent.MEMBER_PENDING,
-            data=member.get_audit_log_data(),
-            organization=org,
-            target_object=org.id,
-            target_user=user,
-        )
+class OrganizationComplianceTask(abc.ABC):
+    """Remove members who don't comply with a new org requirement."""
+
+    log_label = ""
+
+    @abc.abstractmethod
+    def is_compliant(self, member: OrganizationMember) -> bool:
+        """Check whether a member complies with the new requirement."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def call_to_action(self, member: OrganizationMember):
+        """Prompt a member to comply with the new requirement."""
+        raise NotImplementedError()
+
+    def remove_non_compliant_members(self, org_id, actor_id, actor_key_id, ip_address):
+        org = Organization.objects.get_from_cache(id=org_id)
+        actor = User.objects.get(id=actor_id) if actor_id else None
+        actor_key = ApiKey.objects.get(id=actor_key_id) if actor_key_id else None
+
+        def remove_member(member):
+            user = member.user
+            logging_data = {"organization_id": org.id, "user_id": user.id, "member_id": member.id}
+
+            try:
+                member.remove_user()
+                member.save()
+            except (AssertionError, IntegrityError):
+                logger.warning(
+                    f"Could not remove {self.log_label} noncompliant user from org",
+                    extra=logging_data,
+                )
+            else:
+                logger.info(
+                    f"{self.log_label} noncompliant user removed from org", extra=logging_data
+                )
+                AuditLogEntry.objects.create(
+                    actor=actor,
+                    actor_key=actor_key,
+                    ip_address=ip_address,
+                    event=AuditLogEntryEvent.MEMBER_PENDING,
+                    data=member.get_audit_log_data(),
+                    organization=org,
+                    target_object=org.id,
+                    target_user=user,
+                )
+
+                self.call_to_action(member)
+
+        for member in OrganizationMember.objects.select_related("user").filter(
+            organization=org, user__isnull=False
+        ):
+            if not self.is_compliant(org, member.user):
+                remove_member(member)
+
+
+class TwoFactorComplianceTask(OrganizationComplianceTask):
+    log_label = "2FA"
+
+    def is_compliant(self, member: OrganizationMember) -> bool:
```

Comment:
this function takes only member but above we pass it org and member which is causing it to error on my machine

Reply:
is there any way we could add a test to this?

Reply:
Edit: looks like there is a test and it is failing, disregard above

Reply:
Fixed; thanks.